fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,822 @@
+       AIES, 2018.
+
+    """  # noqa : E501
+
+    def __init__(
+        self,
+        *,
+        backend="auto",
+        predictor_model=[],  # [] is a NN with no hidden layers.
+        adversary_model=[],
+        predictor_loss="auto",
+        adversary_loss="auto",
+        predictor_function="auto",
+        predictor_optimizer="Adam",
+        adversary_optimizer="Adam",
+        constraints="demographic_parity",
+        y_transform=FloatTransformer("auto"),
+        z_transform=FloatTransformer("auto"),
+        learning_rate=0.001,
+        alpha=1.0,
+        epochs=1,
+        batch_size=-1,
+        max_iter=-1,
+        shuffle=False,
+        progress_updates=None,
+        skip_validation=False,
+        callbacks=None,
+        cuda=None,
+        warm_start=False,
+        random_state=None,
+    ):
+        """Initialize class by only storing (kw)args, as per sklearn API."""
+        self.backend = backend
+        self.predictor_model = predictor_model
+        self.adversary_model = adversary_model
+        self.predictor_loss = predictor_loss
+        self.adversary_loss = adversary_loss
+        self.predictor_function = predictor_function
+        self.predictor_optimizer = predictor_optimizer
+        self.adversary_optimizer = adversary_optimizer
+        self.constraints = constraints
+        self.y_transform = y_transform
+        self.z_transform = z_transform
+        self.learning_rate = learning_rate
+        self.alpha = alpha
+        self.epochs = epochs
+        self.batch_size = batch_size
+        self.max_iter = max_iter
+        self.shuffle = shuffle
+        self.progress_updates = progress_updates
+        self.skip_validation = skip_validation
+        self.callbacks = callbacks
+        self.cuda = cuda
+        self.warm_start = warm_start
+        self.random_state = random_state
+
+    def __setup(self, X, Y, Z):
+        """
+        Initialize the entire model from the parameters and the given data.
+
+        Following sklearn API, we do not do intialization in `__init__`, but
+        instead in `fit`. Firstly, we validate the backend. Then, we validate
+        some key-word arguments. Then, we initialize the BackendEngine, which
+        handles the initialization of the losses and optimizers. Among these
+        steps, if a loss or function is not explicitely defined, we try to
+        infer something appropriate from data (see interpret_keyword).
+        """
+        self._validate_backend()
+
+        # Verify the constraints and set up the corresponding network structure.
+        if self.constraints == "demographic_parity":
+            self.pass_y_ = False
+        elif self.constraints == "equalized_odds":
+            self.pass_y_ = True
+        else:
+            raise ValueError(
+                _KWARG_ERROR_MESSAGE.format(
+                    "constraints", "'demographic_parity' or 'equalized_odds'"
+                )
+            )
+
+        # Non-negative parameters
```

Comment:
Would the pure validation checks here (as opposed to those which also set internals, such as the constraints check above) be better placed in a separate method?

Reply:
I do agree, but I saw many sklearn methods do it like this

Reply:
@adrinjalali What do you think?